### PR TITLE
[Feature #160577607] Added the user signup payload validation

### PR DESCRIPTION
**Pull request by SrMoffat**
## What does this PR do?
Serves the payload validations for the `api/user` endpoint for *registering* a user

## Pivotal Tracker Story
Users should receive a descriptive error message when they enter invalid user parameters in an attempt to signup.

## Tasks accomplished 

| Endpoint        | Functionality        | Branch  |Pivotal Tracker
| ------------- |-------------| ----------| -------------|
| POST api/users/| REGISTER a user |      `ft-signup-validation-160577607`      |  Story is [here](https://www.pivotaltracker.com/n/projects/2198535/stories/160577607) |

## Manual Testing
`Run server` --> **python manage.py runserver**
`Postman` --> Navigate to the endpoint `api/users/` with a `POST`

## Cases Handled
1. Invalid user parameters handled include:
- Existing user email (**"User already exists. Login!"**)
- Invalid email (**" The email is invalid! Hint: example@mail.com"**)
- Password of length less than 8 (**"Password should be more than 8 characters!"**)
- Password not being alphanumeric (**"Password must have at least one number"**)
- Password not being alphanumeric (**"Password must have at least uppercase letter"**)

2. Additional checks include
- Password not having a special character (**"Password must have at least one special symbol"**)
- Missing username (**"Username is required!"**)
- Missing email (**"Email is required!"**)
- Missing password (**"Password is required!"**)

## Additional comments
The tests for the functionality are coming soon

### CODE REVIEW

**Comment by gitaumoses4** [comment]
Good description on the Pull Request. I like the use of the table. 👍

**Comment by charisschomba** [comment]
How can this be manually tested.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+language: python
```
Comment (esirK) [inline .travis.yml:1]: Is this file related to this task??

File: README.md [hunk 2]
```diff
@@ -1,5 +1,7 @@
+[![Build Status](https://travis-ci.org/andela/ah-code-blooded.svg?branch=develop)](https://travis-ci.org/andela/ah-code-blooded)
```
Comment (esirK) [inline README.md:1]: From what branch did you check out from? This also doesn't seem to be related to this task.

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -1,34 +1,113 @@
...
     # characters, and can not be read by the client.
-    password = serializers.CharField(
+    password = serializers.RegexField(
+        regex="^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$",
         max_length=128,
         min_length=8,
-        write_only=True
+        write_only=True,
+        required=True,
+        error_messages={
```
Comment (esirK) [inline authors/apps/authentication/serializers.py]: This works fine but having a method validation instead of this might help ensure appropriate error messages are returned. e.g **Password MUST have BOTH letters and numbers** should only be returned if the user submits password that doesn't meet this criteria. Checkout this http://www.django-rest-framework.org/api-guide/serializers/#field-level-validation

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -1,34 +1,113 @@
...
     # request. Making `token` read-only handles that for us.
+    token = serializers.CharField(
+        read_only=True
+    )
+    
+    def create(self, validated_data):
+        # Use the `create_user` method we wrote earlier to create a new user.
+        return User.objects.create_user(**validated_data)
+
+    def validate(self, data):
```
Comment (esirK) [inline authors/apps/authentication/serializers.py:63]: This method works but as **Uncle Bob** says in his book on clean architecture, methods should be small. you can resolve this by using **Field-level validation** instead of this. In this case, you can validate every field on its own method.

**Comment by esirK** [comment]
Looks good. Work on the comments and it will be ready for merging.

**Comment by esirK** [comment]
Did you ignore the comments @SrMoffat ???

File: README.md (hunk 2 above)
Comment (SrMoffat) [reply README.md:1]: Given that the branch with the PR was not from develop, I checked out develop and created another branch ft-signup-validators-160577607 to house the code for the PT story. Therefore, this branch should be discarded because it does not use develop as the base. I created a local branch and later created an upstream for it. I have created a new branch that appropriately uses develop as the base. Thanks

File: .travis.yml (hunk 1 above)
Comment (SrMoffat) [reply .travis.yml:1]: As per the comments, the `travis.yml` file was not related to the validations hence I removed it from the new branch I have created to house the validations (`ft-signup-validators-160577607` which houses the requested changes.

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (SrMoffat) [reply authors/apps/authentication/serializers.py:63]: For the identified discrepancy on the README.md, the current version of the file living on the develop branch has two badges for builds and coverage. I believe that is why it should stay to avoid merge conflicts. Finally, I modularised the previously overloaded `validate` function to separate field-based validations: `validate_username` , `validate_email` , and `validate_password`. These changes are reflected in the new branch I created `ft-signup-validators-160577607`
